### PR TITLE
RNN `num_layers` and `dropout` docs and checks

### LINKED ISSUE #5985: RNN `dropout` arg needs better documentation
The documentation for the argument ``dropout`` of the LSTM module [here](http://pytorch.org/docs/master/nn.html#torch.nn.LSTM) confused me. Here's the quote:

>If non-zero, introduces a dropout layer on the outputs of each RNN layer except the last layer

GRU uses the same documentation. Initially I assumed that meant supplying a value other than `False` will implement a dropout layer with a default dropout probability. 

Today I dug through the source code. I went through LSTM, RNN base, Module, THNN backends, Function backends, functions RNN, functions Autograd RNN, and finally [functions Stacked RNN](https://github.com/pytorch/pytorch/blob/master/torch/nn/_functions/rnn.py#L68) before I figured out that the argument is a float probability and not a boolean.

I did some searches on the forums. I'm not the first person to become confused (Exhibit [A](https://discuss.pytorch.org/t/lstm-dropout-clarification-of-last-layer/5588), [B](https://discuss.pytorch.org/t/dropout-in-lstm/7784), [C](https://discuss.pytorch.org/t/dropout-for-rnns/633)). It wasn't until I read through A that I figured out by "layer" it didn't mean the last unwrapped layer but rather the last stacked layer. That is to say, using the arg when `num_layers` is 1 actually did nothing. Possibly worthy of a warning...

I opened a topic on the forums [here](https://discuss.pytorch.org/t/does-rnn-dropout-arg-need-better-documentation/15407). Didn't gain traction, but the one reply at this time suggested an issue or PR.

**Tl;dr: The docs for `nn.LSTM` and `nn.GRU` should say the arg is a float in [0, 1], should clarify what's meant by layer, and maybe the source should issue a warning when `dropout` is not 0 and `num_layers` is 1.**

### PR DESCRIPTION
**Pull request by ssnl**
Improves doc on RNN classes `num_layers` and `dropout` args
Adds value checks for `dropout` arg

Fixes https://github.com/pytorch/pytorch/issues/5985

### CODE REVIEW

File: torch/nn/modules/rnn.py [hunk 1]
```diff
@@ -25,6 +26,16 @@ def __init__(self, mode, input_size, hidden_size,
         self.bidirectional = bidirectional
         num_directions = 2 if bidirectional else 1
 
+        if not isinstance(dropout, numbers.Number) or not 0 <= dropout <= 1:
```
Comment (apaszke) [inline torch/nn/modules/rnn.py]: Why would it be something else than a number

File: torch/nn/modules/rnn.py [hunk 2]
```diff
@@ -25,6 +26,16 @@ def __init__(self, mode, input_size, hidden_size,
...
+            raise ValueError("dropout should be a number in range [0, 1] "
+                             "representing the probablity of an element being "
+                             "zeroed")
+        if dropout > 0 and num_layers == 1:
+            raise ValueError("dropout option adds dropout after all but last "
```
Comment (apaszke) [inline torch/nn/modules/rnn.py]: I feel like it's ok to make this one a warning (e.g. for debugging?). wdyt?

File: torch/nn/modules/rnn.py (hunk 1 above)
Comment (SsnL) [reply torch/nn/modules/rnn.py]: Some people were confused by doc and put a bool here.

File: torch/nn/modules/rnn.py (hunk 2 above)
Comment (SsnL) [reply torch/nn/modules/rnn.py]: Sure.

File: torch/nn/modules/rnn.py (hunk 1 above)
Comment (apaszke) [reply torch/nn/modules/rnn.py]: Isn’t Bool a Number too? IIRC it’s a subclass

Comment (SsnL) [reply torch/nn/modules/rnn.py]: You are absolutely correct. I added a check for bool.

File: torch/nn/modules/rnn.py [hunk 3]
```diff
@@ -25,6 +26,17 @@ def __init__(self, mode, input_size, hidden_size,
         self.bidirectional = bidirectional
         num_directions = 2 if bidirectional else 1
 
+        if not isinstance(dropout, numbers.Number) or not 0 <= dropout <= 1 or \
+                isinstance(dropout, bool):
```
Comment (ezyang) [inline torch/nn/modules/rnn.py]: lol, I didn't know bool is an instance of Number

**Review by ezyang — APPROVED** [review]

**Review by apaszke — APPROVED** [review]

File: torch/nn/modules/rnn.py (hunk 3 above)
Comment (jekbradbury) [reply torch/nn/modules/rnn.py]: in fact `bool`s are instances of `int`!
